plugins/microservices/msi_json/json_parse.cpp

```diff
@@ -0,0 +1,148 @@
+/// \file
+
+#include "msi_preconditions.hpp"
+
+#include "irods_error.hpp"
+#include "irods_ms_plugin.hpp"
+#include "irods_re_structs.hpp"
+#include "msParam.h"
+#include "objInfo.h"
+#include "process_stash.hpp"
+#include "rcMisc.h"
+#include "rodsErrorTable.h"
+#include "rodsPackInstruct.h"
+#include "rodsLog.h"
+
+#include <fmt/format.h>
+#include <json.hpp>
+
+#include <algorithm>
+#include <functional>
+#include <iterator>
+#include <string>
+
+namespace
+{
+    using json = nlohmann::json;
+
+    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    auto msi_impl(MsParam* _range, MsParam* _range_size, MsParam* _json_handle, ruleExecInfo_t* _rei) -> int
+    {
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range_size);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_json_handle);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range->type);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range_size->type);
+
+        IRODS_MSI_REQUIRE_TYPE(_range->type, STR_MS_T, IntArray_MS_T);
+        IRODS_MSI_REQUIRE_TYPE(_range_size->type, INT_MS_T);
+
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range->inOutStruct);
+        IRODS_MSI_REQUIRE_VALID_POINTER(_range_size->inOutStruct);
+
+        try {
+            const auto buf_size = *static_cast<int*>(_range_size->inOutStruct);
+
+            if (buf_size <= 0) {
+                rodsLog(LOG_ERROR, "Invalid argument: Buffer size must be greater than zero.");
+                return SYS_INVALID_INPUT_PARAM;
+            }
+
+            std::string handle;
+
+            if (const std::string_view t = _range->type; t == STR_MS_T) {
+                const auto* char_array = static_cast<const char*>(_range->inOutStruct);
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+                handle = irods::process_stash::insert(json::parse(char_array, char_array + buf_size));
+            }
+            else if (t == IntArray_MS_T) {
+                const auto* int_array = static_cast<IntArray*>(_range->inOutStruct);
+
+                if (buf_size > int_array->len) {
+                    constexpr const char* msg = "Invalid argument: Buffer size [%i] exceeds size of integer sequence [%i].";
+                    rodsLog(LOG_ERROR, msg, buf_size, int_array->len);
+                    return SYS_INVALID_INPUT_PARAM;
+                }
+
+                std::vector<char> buf;
+                buf.reserve(buf_size);
+
+                std::transform(int_array->value, int_array->value + buf_size, std::back_inserter(buf), [](auto _v) {
+                    return static_cast<char>(_v);
+                });
+
+                handle = irods::process_stash::insert(json::parse(std::begin(buf), std::end(buf)));
```

Comment:
Here's a significant change from main.

4-2-stable uses a version of the JSON lib that requires that all elements of a container you wish to parse have a size of 1 byte and be contiguous. Because of that, we are required to **copy** the range if it represents an array containing integers.

Ultimately, integer arrays consume double the amount of memory when this microservice is invoked.

I'm going to add a note to the documentation about that.

Reply:
Makes sense.

Reply:
Done.